### DISCUSSION TITLE
any guide to migrate from 1.x to 2.x?

### DISCUSSION THREAD

**Discussion by sombriks** [brgl/libgpiod#56] [Q&A] [answered]
i am trying to proper branch my node wrapper and  maintain 1.x and 2.x, but i am figuring a few issues on new names.

can anyone point me where could i read  how to get the new versions of those functions:

chip:

gpiod_chip_open_lookup => gpiod_chip_open
gpiod_chip_num_lines => gpiod_chip_info_get_num_lines
gpiod_chip_name => gpiod_chip_info_get_name
gpiod_chip_label => gpiod_chip_info_get_label

line:

gpiod_chip_get_line => gpiod_chip_get_line_info
gpiod_line_close_chip => gpiod_line_info_free
gpiod_line_offset = > gpiod_line_info_get_offset
gpiod_line_name => gpiod_line_info_get_name
gpiod_line_consumer => gpiod_line_info_get_consumer
gpiod_line_get_value => ???
gpiod_line_set_value => ???
gpiod_line_request_input => ???
gpiod_line_request_input_flags => ???
gpiod_line_request_output => ???
gpiod_line_release => ???

**Comment by brgl** [comment, marked as answer]
It'll be quite hard to create a wrapper for both as the API has been completely rebuilt and there are no 1:1 translations between the interfaces. You no longer *get* the line and then *request* it. You know request a set of lines and receive a `gpiod_line_request` object that can wrap multiple lines.

**Reply by sombriks** [reply]
Got it @brgl , thanks for the guidance!

So indeed i must maintain one thing for 1.x and the new one for 2.x

**Comment by mar0x** [comment]
@sombriks Have you found a solution?
@brgl Is there compile-time version macro to use conditional compilation?

**Reply by sombriks** [reply]
hello @mar0x ,

on [my node wrapper](https://github.com/sombriks/node-libgpiod) i'll offer two versions, 1.x and 2.x (and maybe preserve 0.x for pre 1.5 versions or older), each one offering the distinct api.

right now i am working on make 1.x feature-parity complete, then i'll start to work on 2.x. using my spare time, so the progress is slow but there is progress.

**Reply by mar0x** [reply]
Hello @sombriks ,

Thanks for reply!
My project is in C and it seems I will have to re-package 1.6.x and solve the issue on packaging level.

**Comment by raffaeler** [comment]
On my side, it's ok to create a different interop library (C#), but I cannot find any info on:
1. What is the version described in the official docs? https://libgpiod.readthedocs.io/en/latest/core_api.html I see no version number
2. Is there any doc explaining the sequence of calls to build the new interop layer?

Since the old ABI suddenly disappeared in Trixie, the community is totally disrupted now. It would be nice to clearly document what to do.
Thanks

**Reply by brgl** [reply]
The version in latest is the master branch of this repo, though I should see about displaying it in the docs explicitly. I don't remember if there was an option for that in RTD.

Does "interop" mean the same thing as "bindings"? If so then you could use the python or C++ bindings as example I suppose?

**Reply by raffaeler** [reply]
> The version in latest is the master branch of this repo, though I should see about displaying it in the docs explicitly. I don't remember if there was an option for that in RTD.

That would be great, thanks

> Does "interop" mean the same thing as "bindings"? If so then you could use the python or C++ bindings as example I suppose?

Not only. I mean  that beyond wrapping the calls to invoke C functions, I expect a sort of flowchart to govern the state of the operation. For example: open, read, close. I expect more complex state for events and other operations.

TIA
